### ISSUE TITLE
Complex matrices failing to convert datatype

### ISSUE THREAD

**Issue by uncommoncode** [wouterboomsma/eigency#19]
Hi, it seems that complex matrices may be broken with the current master. Here is an example:

`test.pyx`:

```
from eigency.core cimport *

cdef extern from "test.h":
     cdef MatrixXcd _test_matrix "test_matrix"(Map[MatrixXcd] &)

def test_matrix(np.ndarray array):
    return ndarray_copy(_test_matrix(Map[MatrixXcd](array)))
```

`test.h`:

```
#pragma once

#include <eigen3/Eigen/Dense>

Eigen::MatrixXcd test_matrix(const Eigen::MatrixXcd&);
```

`test.cpp`:

```
#include "test.h"

Eigen::MatrixXcd test_matrix(const Eigen::MatrixXcd& result) {
    return result;
}
```

Then in python:

```
import test_matrix
print test_matrix.test_matrix(np.zeros((3, 3)) + 0.4)
```

The output is not as expected (`0.4 + 0.0j` for all entries):

```
[[  4.00000000e-001 +4.00000000e-001j   4.00000000e-001 +4.00000000e-001j
    6.93540132e-310 +6.93540132e-310j]
 [  4.00000000e-001 +4.00000000e-001j   4.00000000e-001 +6.95335581e-309j
    0.00000000e+000 +6.93540132e-310j]
 [  4.00000000e-001 +4.00000000e-001j   6.93540132e-310 +0.00000000e+000j
    6.93540132e-310 +0.00000000e+000j]]
```

However, if I do something like the following:

```
m = np.zeros((3, 3)) + 0.4
print test_matrix.test_matrix(m.astype(np.complex128))
```

I get the correct output:

```
[[ 0.4+0.j  0.4+0.j  0.4+0.j]
 [ 0.4+0.j  0.4+0.j  0.4+0.j]
 [ 0.4+0.j  0.4+0.j  0.4+0.j]]
```

While this example shows mixing real with complex types, it also happens between different complex types, like `complex64` mixed with `complex128`.

Is this expected behavior?

**Comment by wouterboomsma** [comment]
Hi,

I think this indeed falls under the category "intentional", although I'm open to suggestions if there are better ways to deal with this. 

The basic idea behind eigency is to share the underlying representation of a numpy array between Python and C++. This means that somewhere in the process, we need to make explicit which numerical types we are dealing with. In your case, you are passing in an array with float64, and then - from the point of eigency - explicitly asking it to interpret the buffer as a complex128, which will give you the result you describe.

There are various ways around this:

1. You can explicitly specify the type that you expect in the cython definition:
```
 def test_matrix(np.ndarray[np.complex128_t, ndim=2] array):   # instead of just np.ndarray
    return ndarray_copy(_test_matrix(Map[MatrixXcd](array)))
````
Doing so will raise an error if someone passes something that is not of the expected type.

2. Alternatively, you can of course also just do the conversion you describe within cython:
```
def test_matrix(np.ndarray array):
    return ndarray_copy(_test_matrix(Map[MatrixXcd](array.astype(np.complex128))))
```

You could argue that calling Map[MatrixXcd] on an array of type float64 should be illegal. As far as I can see, that would however require us to implement template specializations for every type of Map object, where it for each type is hardcoded which python dtypes are allowed. This is in principle doable, but it would mostly be useful for providing an error, but this is already supported by option 1 above. Of course, the template specializations could in principle do the appropriate conversion to a different type automatically - but that would have surprising effects for people who expect that the memory storage is shared between Python and C++ (which it would not be if the storage type changes behind the scenes). 
Long story short: In Eigency we leave it up to the designer of the interface to specify how specific the interface should be. You can should yourself in the foot this way - but c++ programmers are used to that :)

**Comment by uncommoncode** [comment]
Thanks this is very helpful! I wasn't aware of the extra type info you can put on the numpy array like `np.ndarray[np.complex128_t, ndim=2]`. Seems like the way to go :D

Perhaps it would be useful to have this in the otherwise stellar readme?

**Comment by wouterboomsma** [comment]
Sorry that I never got back to you on this. It was on my todo list but I didn't get around to it until today. I decided to make the explicit typing of numpy arrays the recommended style, and have now updated both the README.md with a section on this (Types matter) - and modified all test code examples so that they now all explicitly state which numpy array type they expect as input.
